File: examples/agent-awareness.py
"""
Floe Agent Awareness — Decision-loop demo.

Answers the three rational-agent questions before paying for an API call:
  1. Do I have enough credit?
  2. Is this call worth its cost?
  3. Where am I in the loan lifecycle?

Usage:
  FLOE_API_KEY=floe_... TARGET_URL=https://some-x402-api.com/data \
    python agent-awareness.py

Mirror of agent-awareness.ts — same behavior, same output format.
"""

from __future__ import annotations

import json
import os
import sys
import urllib.error
import urllib.request


API_KEY = os.environ.get("FLOE_API_KEY")
if not API_KEY:
    sys.stderr.write("Set FLOE_API_KEY=floe_... environment variable\n")
    sys.exit(1)

BASE = os.environ.get("FLOE_API_BASE", "https://credit-api.floelabs.xyz")
TARGET_URL = os.environ.get("TARGET_URL", "https://some-x402-api.com/data")

HEADERS = {
    "Authorization": f"Bearer {API_KEY}",
    "Content-Type": "application/json",
}


def get(path: str) -> dict:
    req = urllib.request.Request(f"{BASE}{path}", headers=HEADERS, method="GET")
    with urllib.request.urlopen(req, timeout=30) as resp:
        return json.loads(resp.read())


def post(path: str, body: dict) -> dict:
    req = urllib.request.Request(
        f"{BASE}{path}",
        data=json.dumps(body).encode(),
        headers=HEADERS,
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=30) as resp:
        return json.loads(resp.read())


def main() -> None:
    # Question 3 first — gate everything on loan state
    state = get("/v1/agents/loan-state")
    print(f"\n[loan-state] {state['state']} — {state.get('reason', '')}")
    if state["state"] == "at_limit":
        print("Skipping: agent is at_limit. Open another credit line or repay.")
        return
    if state["state"] in ("borrowing", "repaying"):
        print("Skipping: pending capital movement. Try again shortly.")
        return

    # Question 2 — estimate cost AND reflect against agent state in one call
    est = post("/v1/x402/estimate", {"url": TARGET_URL, "method": "GET"})
    print(f"\n[estimate] {est['method']} {est['url']}")
    if not est.get("x402"):
        print("URL is not x402-protected — call it directly.")
        return
    print(f"  price: {est.get('priceRaw')} raw USDC ({est.get('network')})")
    print(f"  cached: {est.get('cached')}")

    # Fail-closed: if the API didn't return a reflection block, abort. The
    # reflection is what tells us whether the call would exceed available
    # credit or our session spend-limit; without it we have no decision basis,
    # so don't fall through to the paid call.
    r = est.get("reflection")
    if not r:
        print("  ❌ estimate response missing reflection block — DO NOT CALL")
        return
    print(f"  available: {r.get('available')}")
    if r.get("willExceedAvailable"):
        print("  ❌ would exceed available credit — DO NOT CALL")
        return
    if r.get("willExceedSpendLimit"):
        print("  ❌ would exceed session spend-limit — DO NOT CALL")
        return
    if r.get("willExceedHeadroom"):
        print("  ⚠️  would dip into auto-borrow headroom — proceeding (informational)")

    # Question 1 was answered by the reflection block above. Proceed.
    print("\n[proxy/fetch] paying and calling target...")
    paid = post("/v1/proxy/fetch", {"url": TARGET_URL, "method": "GET"})
    print("  response:", json.dumps(paid)[:200])

    # After-call sanity check
    remaining = get("/v1/agents/credit-remaining")
    print(
        f"\n[after] available: {remaining['available']}, "
        f"utilization: {remaining['utilizationBps']} bps"
    )


if __name__ == "__main__":
    try:
        main()
    except urllib.error.HTTPError as e:
        sys.stderr.write(f"HTTP {e.code}: {e.read().decode(errors='replace')}\n")
        sys.exit(1)
    except Exception as e:
        sys.stderr.write(f"Error: {e}\n")
        sys.exit(1)

File: examples/agent-awareness.ts
/**
 * Floe Agent Awareness — Decision-loop demo.
 *
 * Answers the three rational-agent questions before paying for an API call:
 *   1. Do I have enough credit?
 *   2. Is this call worth its cost?
 *   3. Where am I in the loan lifecycle?
 *
 * Usage:
 *   FLOE_API_KEY=floe_... TARGET_URL=https://some-x402-api.com/data \
 *     npx tsx agent-awareness.ts
 *
 * Mirror of agent-awareness.py — same behavior, same output format.
 */

const API_KEY = process.env.FLOE_API_KEY;
if (!API_KEY) {
  console.error("Set FLOE_API_KEY=floe_... environment variable");
  process.exit(1);
}

const BASE = process.env.FLOE_API_BASE || "https://credit-api.floelabs.xyz";
const TARGET_URL = process.env.TARGET_URL || "https://some-x402-api.com/data";

const headers = {
  Authorization: `Bearer ${API_KEY}`,
  "Content-Type": "application/json",
};

async function get(path: string): Promise<any> {
  const r = await fetch(`${BASE}${path}`, { headers });
  if (!r.ok) throw new Error(`GET ${path} → ${r.status} ${await r.text()}`);
  return r.json();
}

async function post(path: string, body: unknown): Promise<any> {
  const r = await fetch(`${BASE}${path}`, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error(`POST ${path} → ${r.status} ${await r.text()}`);
  return r.json();
}

async function main(): Promise<void> {
  // Question 3 first — gate everything on loan state
  const state = await get("/v1/agents/loan-state");
  console.log(`\n[loan-state] ${state.state} — ${state.reason}`);
  if (state.state === "at_limit") {
    console.log("Skipping: agent is at_limit. Open another credit line or repay.");
    return;
  }
  if (state.state === "borrowing" || state.state === "repaying") {
    console.log("Skipping: pending capital movement. Try again shortly.");
    return;
  }

  // Question 2 — estimate cost AND reflect against agent state in one call
  const est = await post("/v1/x402/estimate", { url: TARGET_URL, method: "GET" });
  console.log(`\n[estimate] ${est.method} ${est.url}`);
  if (!est.x402) {
    console.log("URL is not x402-protected — call it directly.");
    return;
  }
  console.log(`  price: ${est.priceRaw} raw USDC (${est.network})`);
  console.log(`  cached: ${est.cached}`);

  const r = est.reflection;
  if (r) {
    console.log(`  available: ${r.available}`);
    if (r.willExceedAvailable) {
      console.log("  ❌ would exceed available credit — DO NOT CALL");
      return;
    }
    if (r.willExceedSpendLimit) {
      console.log("  ❌ would exceed session spend-limit — DO NOT CALL");
      return;
    }
    if (r.willExceedHeadroom) {
      console.log("  ⚠️  would dip into auto-borrow headroom — proceeding (informational)");
    }
  }

  // Question 1 was answered by the reflection block above. Proceed.
  console.log("\n[proxy/fetch] paying and calling target...");
  const paid = await post("/v1/proxy/fetch", { url: TARGET_URL, method: "GET" });
  console.log("  response:", JSON.stringify(paid).slice(0, 200));

  // After-call sanity check
  const remaining = await get("/v1/agents/credit-remaining");
  console.log(`\n[after] available: ${remaining.available}, utilization: ${remaining.utilizationBps} bps`);
}

main().catch((e) => {
  console.error("Error:", e instanceof Error ? e.message : e);
  process.exit(1);
});
